### ISSUE TITLE
Export Cmake imported targets even for header-only components

### ISSUE THREAD

**Issue by aslze** [PointCloudLibrary/pcl#3397]
PCLConfig.cmake, as of PCL 1.9 (AFAIK) added generation of IMPORTED targets for PCL components. That is a good move. However, the script skips header-only components (such as geometry).

Header-only components can create imported targets in the form of INTERFACE libraries.

```cmake
add_library(pcl_geometry INTERFACE IMPORTED)
set_target_properties(pcl_geometry PROPERTIES
	INTERFACE_INCLUDE_DIRECTORIES "${PCL_GEOMETRY_INCLUDE_DIRS}")
```

This way all components would use a consistent way of importing and using them:

```cmake
find_package(PCL COMPONENTS common geometry io)

target_link_libraries(myapp pcl_common pcl_geometry pcl_io)
```

**Comment by kunaltyagi** [comment]
Oops, switched the last 2 digits while referencing!! Sorry.

@aslze I took a glance at the current functions in CMake used to declare the library and it seems that it'll take a non trivial effort to do this properly if we go by updating the `PCL_ADD_LIBRARY` [function](https://github.com/PointCloudLibrary/pcl/blob/master/cmake/pcl_targets.cmake#L216) at least. The type of library (STATIC or SHARED) is common to the entire build which makes it a tad bit difficult to do.

Adding a `PCL_ADD_HEADER_ONLY_LIBRARY` or something less mouthful might be a good alternative. Please feel free to create a PR

**Comment by aslze** [comment]
I was refering to the `PCLConfig.cmake.in` file, not the scripts used when creating the libraries (each of the modules). In that file, at line 590, there is a condition:

```cmake
   if(_is_header_only EQUAL -1)   # so, if this component is NOT header-only
```
followed by `add_library(... IMPORTED)` and `set_target_properties` to populate its properties (include dirs, location, defines, etc.).

But that `if()` has no  `else()`, so no imported targets are created for header-only libs.

I thought about adding an `else` block with something like:

```cmake
    else() # header-only
      add_library(${pcl_component} INTERFACE IMPORTED)
        set_target_properties(${pcl_component} PROPERTIES
            INTERFACE_COMPILE_DEFINITIONS "${definitions}"
            INTERFACE_COMPILE_OPTIONS "${PCL_COMPILE_OPTIONS}" # Needed?
            INTERFACE_INCLUDE_DIRECTORIES "${PCL_${COMPONENT}_INCLUDE_DIRS};${PCL_CONF_INCLUDE_DIR}"
        )
    endif()
```

As a workaround (before I try to modify anything in PCL) I am doing this externally, in a script that calls `find_package(PCL)` and then builds the imported targets from the generated `PCL_*` variables. It kind of works but is still incomplete.

And, yes, I am seeing that doing it properly is *not at all trivial*. Especially dealing with internal and external dependencies which have to be added, and avoiding those having absolute paths, etc.

**Comment by taketwo** [comment]
Good catch, an `else` block along the lines of what you posted is indeed missing. A patch is welcome.

**Comment by kunaltyagi** [comment]
Resolved by #3495
